tutorials/roofit/rf317_global_observables.C

```diff
@@ -0,0 +1,169 @@
+   // reformulate the likelihood to make it apparent that `lumi_obs` is a
+   // global observables:
+   //
+   // l'(data | mu, lumi) = l(data | mu, lumi) * Gauss(lumi_obs | lumi, lumi_obs_sigma)
+   //
+   // Note: be extra careful to define global observables and constrained
+   // parameters correctly in the case of non-symmetric constraints pdfs! For
+   // example, if the constraint would be a Poissonian, the global observable
+   // really has to be the observable of the Poissonian:
+   //
+   // l'(data | mu, lumi) = l(data | mu, lumi) * Poisson(lumi_obs | lumi)
+   //
+
+   // Setting up the model and creating toy dataset
+   // ---------------------------------------------
+
+   // l'(x | mu, sigma) = l(x | mu, sigma) * Gauss(mu_obs | mu, 0.2)
+
+   // event observables
+   RooRealVar x("x", "x", -10, 10);
+
+   // parameters
+   RooRealVar mu("mu", "mu", 0.0, -10, 10);
+   RooRealVar sigma("sigma", "sigma", 1.0, 0.1, 2.0);
+
+   // Gaussian model for event observables
+   RooGaussian gauss("gauss", "gauss", x, mu, sigma);
+
+   // global observables (which are no parameters so they are constant)
+   RooRealVar mu_obs("mu_obs", "mu_obs", 1.0, -10, 10);
+   mu_obs.setConstant();
+
+   // constraint pdf
+   RooGaussian constraint("constraint", "constraint", mu_obs, mu, RooConst(0.2));
+
+   // full pdf including constraint pdf
+   RooProdPdf model("model", "model", {gauss, constraint});
+
+   // Generating toy data with randomized global observables
+   // ------------------------------------------------------
+
+   // For most toy-based statistical procedures, it is necessary to also
+   // randomize the global observable when generating toy datasets.
+   //
+   // To that end, let's generate a single event from the model and take the
+   // global observable value (the same is done in the RooStats:ToyMCSampler
+   // class):
+
+   std::unique_ptr<RooDataSet> dataGlob{model.generate({mu_obs}, 1)};
+
+   // Next, we temporarily set the value of `mu_obs` to the randomized value for
+   // generating our toy dataset:
+   double mu_obs_orig_val = mu_obs.getVal();
+
+   RooArgSet{mu_obs}.assign(*dataGlob->get(0));
+
+   // actually generate the toy dataset
+   std::unique_ptr<RooDataSet> data{model.generate({x}, 1000)};
+
+   // When fitting the toy dataset, it is important to set the global
+   // observables in the fit to the values that were used to generate the toy
+   // dataset. To facilitate the bookkeeping of global observable values, you
+   // can attach a snapshot with the current global observable values to the
+   // dataset like this (new feature introduced in ROOT 6.26):
+
+   data->setGlobalObservables({mu_obs});
+
+   // reset original mu_obs value
+   mu_obs.setVal(mu_obs_orig_val);
+
+   // Note that in a future ROOT release, we plan to streamline the sampling of
+   // toy datasets with randomized global observables such that it can be done
+   // with one call to RooAbsPdf::generate():
```

Comment:
Those kind of comments degrade fast. If you forget to remove this, it looks bad. If you don't implement this, it also looks bad. 🙂

Reply:
True! I removed it.